icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java

```diff
@@ -693,8 +770,458 @@ public void unitCompoundMeasure() {
                 "0.08765 J/fur",
                 "0.008765 J/fur",
                 "0 J/fur");
+
+        // TODO(icu-units#35): does not normalize as desired: while "unit/*" does
+        // get split into unit/perUnit, ".unit(*)" and "measure-unit/*" don't:
+        assertFormatSingle(
+                "Built-in unit, meter-per-second",
+                "measure-unit/speed-meter-per-second",
+                "~unit/meter-per-second",
+                NumberFormatter.with().unit(MeasureUnit.METER_PER_SECOND),
+                new ULocale("en-GB"),
+                2.4,
+                "2.4 m/s");
+
+        // TODO(icu-units#59): THIS UNIT TEST DEMONSTRATES UNDESIRABLE BEHAVIOUR!
+        // When specifying built-in types, one can give both a unit and a perUnit.
+        // Resolving to a built-in unit does not always work.
+        //
+        // (Unit-testing philosophy: do we leave this enabled to demonstrate current
+        // behaviour, and changing behaviour in the future? Or comment it out to
+        // avoid asserting this is "correct"?)
+        assertFormatSingle(
+                "DEMONSTRATING BAD BEHAVIOUR, TODO(icu-units#59)",
+                "measure-unit/speed-meter-per-second per-measure-unit/duration-second",
+                "measure-unit/speed-meter-per-second per-measure-unit/duration-second",
+                NumberFormatter.with()
+                        .unit(MeasureUnit.METER_PER_SECOND)
+                        .perUnit(MeasureUnit.SECOND),
+                new ULocale("en-GB"),
+                2.4,
+                "2.4 m/s/s");
+
+        // Testing the rejection of invalid specifications
+
+        // If .unit() is not given a built-in type, .perUnit() is not allowed
+        // (because .unit is now flexible enough to handle compound units,
+        // .perUnit() is supported for backward compatibility).
+        LocalizedNumberFormatter nf = NumberFormatter.with()
+                .unit(MeasureUnit.forIdentifier("furlong-pascal"))
+                .perUnit(MeasureUnit.METER)
+                .locale(new ULocale("en-GB"));
+
+        try {
+            nf.format(2.4d);
```

Comment:
This failure looks a bit painful, because LongNameHandler.java:198 is not part of this PR:

java.lang.NullPointerException
at com.ibm.icu.util.MeasureUnit.hashCode(MeasureUnit.java:578)
at com.ibm.icu.impl.Pair.hashCode(Pair.java:53)
at java.base/java.util.HashMap.hash(HashMap.java:339)
at java.base/java.util.HashMap.get(HashMap.java:552)
at com.ibm.icu.util.MeasureUnit.resolveUnitPerUnit(MeasureUnit.java:730)
**at com.ibm.icu.impl.number.LongNameHandler.forMeasureUnit(LongNameHandler.java:198)**
at com.ibm.icu.number.NumberFormatterImpl.macrosToMicroGenerator(NumberFormatterImpl.java:365)
at com.ibm.icu.number.NumberFormatterImpl.preProcessUnsafe(NumberFormatterImpl.java:94)
at com.ibm.icu.number.NumberFormatterImpl.formatStatic(NumberFormatterImpl.java:70)
at com.ibm.icu.number.LocalizedNumberFormatter.formatImpl(LocalizedNumberFormatter.java:159)
at com.ibm.icu.number.LocalizedNumberFormatter.format(LocalizedNumberFormatter.java:133)
at com.ibm.icu.number.LocalizedNumberFormatter.format(LocalizedNumberFormatter.java:71)
at com.ibm.icu.dev.test.number.NumberFormatterApiTest.unitCompoundMeasure(NumberFormatterApiTest.java:814)

This is an invalid/unsupported pair of "unit().perUnit()" calls. In C++ code I provide the error like this:

    if (uprv_strlen(unitRef.getType()) == 0 || uprv_strlen(perUnit.getType()) == 0) {
        // TODO(ICU-20941): Unsanctioned unit. Not yet fully supported. Set an
        // error code. Once we support not-built-in units here, unitRef may be
        // anything, but if not built-in, perUnit has to be "none".
        status = U_UNSUPPORTED_ERROR;
        return;
    }

I think this check can be added in LongNameHandler forMeasureUnit, avoiding needing to deal with MeasureUnit.java further right now.

This check predates my commits. (7ed2a2d23383658f6d52a8514c0c9708fc56b44a and e3123c83a4cdb73458e38f864bda5d5409308e8d touch code nearby, but the checks were in place.) My unit tests just manage to trigger this case. It was Shane's fix in March: fc6612cc56fa1e6ae0dccc55451e6a686085aab3 "ICU-20920 Add support for CLDR 37 unit identifiers in ICU4C".

Reply:
I've fixed this NullPointerException with e2ed40d.

Reply:
done